python/mspasspy/global_history/manager.py

```diff
@@ -1,21 +1,179 @@
+                wf,
+                *args,
+                object_history=object_history,
+                alg_name=alg_name,
+                alg_id=str(alg_id),
+                **kwargs
+            ),
+            data,
+        )
+
+    return map(
+        lambda wf: func(wf, *args, object_history=object_history, **kwargs), data
+    )
+
+
+def mspass_normal_reduce(
+    data,
+    func,
+    *args,
+    global_history=None,
+    object_history=False,
+    alg_id=None,
+    alg_name=None,
+    parameters=None,
+    **kwargs
+):
+    """
+     This method performs reduce function using functools. Instead of performing the normal reduce function,
+     if user provides global history manager, alg_id(optional), alg_name(optional) and parameters(optional)
+     as input, the global history manager will log down the usage of the algorithm. Also, if user set
+     object_history to be True, then each mspass object in this reduce function will save the object level history.
+
+    :param data: data to be processed
+    :param func: target function
+    :param global_history: a user specified global history manager
+    :type global_history: :class:`GlobalHistoryManager`
+    :param object_history: save the each object's history in the reduce when True
+    :param alg_id: a user specified alg_id for the reduce operation
+    :type alg_id: :class:`str`/:class:`bson.objectid.ObjectId`
+    :param alg_name: a user specified alg_name for the reduce operation
+    :type alg_name: :class:`str`
+    :param parameters: a user specified parameters for the reduce operation
+    :type parameters: :class:`str`
+    :return: reduced objects.
+    """
+    if parameters:
+        parameterGTree = parameter_to_GTree(parameters_str=parameters)
+    else:
+        new_kwargs = kwargs.copy()
+        new_kwargs["object_history"] = object_history
+        if alg_name:
+            new_kwargs["alg_name"] = alg_name
+        if alg_id:
+            new_kwargs["alg_id"] = alg_id
+        parameterGTree = parameter_to_GTree(*args, **new_kwargs)
+
+    parameters_json = json.dumps(parameterGTree.asdict())
+
+    if not alg_name:
+        alg_name = func.__name__
+
+    # get the alg_id if exists, else create a new one
+    if not alg_id:
+        # get the alg_id if exists
+        alg_id = global_history.get_alg_id(alg_name, parameters_json)
+        # else create a new one
+        if not alg_id:
+            alg_id = ObjectId()
+
+    # save the global history
+    if global_history:
+        global_history.logging(alg_id, alg_name, parameters_json)
+
+    # save the object history
+    if object_history:
+        return functools.reduce(
+            lambda a, b: func(
+                a,
+                b,
+                *args,
+                object_history=object_history,
+                alg_name=alg_name,
+                alg_id=str(alg_id),
+                **kwargs
+            ),
+            data,
+        )
+
+    return functools.reduce(
```

Comment:
Is this gonna work? The data here needs to be a iterable, right? This is not mentioned in the docstrings. `args` is not needed for `functools.reduce`, right?